### ISSUE TITLE
Error "_set_blas_variables: blas_lib not set" in EasyBuild 3.9.1

### ISSUE THREAD

**Issue by teojgo** [easybuilders/easybuild-framework#2911]
I am getting the following error when I am trying to build LAMMPS with EasyBuild 3.9.1. 
For an extended dry run, the following is included in the logs:
```
WARNING: ignoring error '_set_blas_variables: BLAS_LIB not set'
```

Using EasyBuild 3.8.1 the build succeeds. The eb recipe is this https://github.com/eth-cscs/production/blob/master/easybuild/easyconfigs/l/LAMMPS/LAMMPS-22Aug2018-CrayGNU-18.08.eb,

**Comment by gppezzi** [comment]
You can find [here](https://gist.github.com/gppezzi/6108ee6ff0e3933b33c0ddd946e9424e) a complete error log.

This build fails in the second build `omp` and has the following `buildopts`:
```
buildopts = [ ' mpi ', ' omp ' ]
```
The build seems to work if we build only `omp`:
```
buildopts = [  ' omp ' ]
```

**Comment by ocaisa** [comment]
This really sounds like an error related to the introduction of `multi_deps`, I suspect there is some resetting done related to that that is kicking the error. I will dig into it more today.

**Comment by ocaisa** [comment]
Ok, my gut tells me it is this: https://github.com/easybuilders/easybuild-framework/pull/2810/files#diff-a1bcbbc1aab6b59688b219f298fddb56R1858
I'll see if I can come up with a solution

**Comment by ocaisa** [comment]
It took me a while but I can reproduce this problem, I can fix it by overriding some additional internals of the `LinAlg` class in the `LibSci` class:
```
    def _set_blas_variables(self):
        """Skip setting BLAS related variables"""
        pass

    def _set_lapack_variables(self):
        """Skip setting LAPACK related variables"""
        pass
```
This is unlikely to have any breaking impact since there is no way these were being called in the past (as written they would have always failed for Cray).

Having said all that, I cannot find exactly where/why these functions are being called. It is almost certainly to do with the iteration counting that was introduced in `multi_deps` but I cannot source it.

How to proceed @boegel? I will open a PR with the proposed changes but I'm not 100% happy that I couldn't source why this is now being called.

**Comment by ocaisa** [comment]
The root cause of this is that somewhere between the first and second iteration of the build, the `BLAS_LIB` constant is being cleaned out of the `LibSci` class.

**Comment by ocaisa** [comment]
Ok, the fix is there in #2914 , I admit to not fully understanding why this is necessary, is there some kind of garbage collection during the `reset()`?

The change is trivial though and backwards compatible.

**Comment by boegel** [comment]
Some more background on this...

Full traceback for the bug, as produced by the test added in #2914 (without the accompanying fix):

```
$ python -O -m test.framework.toolchain cray_reset
INFO: This is (based on) vsc.install.shared_setup 0.12.7
Filtered ToolchainTest tests using 'cray_reset', retained 1/43 tests: test_cray_reset
E
======================================================================
ERROR: test_cray_reset (__main__.ToolchainTest)
Test toolchain preparation after reset for Cray* toolchain.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Volumes/work/easybuild-framework/test/framework/toolchain.py", line 191, in test_cray_reset
    tc.prepare()
  File "/Volumes/work/easybuild-framework/easybuild/toolchains/compiler/craype.py", line 125, in prepare
    super(CrayPECompiler, self).prepare(*args, **kwargs)
  File "/Volumes/work/easybuild-framework/easybuild/tools/toolchain/toolchain.py", line 735, in prepare
    self.set_variables()
  File "/Volumes/work/easybuild-framework/easybuild/tools/toolchain/compiler.py", line 164, in set_variables
    super(Compiler, self).set_variables()
  File "/Volumes/work/easybuild-framework/easybuild/tools/toolchain/mpi.py", line 100, in set_variables
    super(Mpi, self).set_variables()
  File "/Volumes/work/easybuild-framework/easybuild/tools/toolchain/linalg.py", line 91, in set_variables
    self._set_blas_variables()
  File "/Volumes/work/easybuild-framework/easybuild/tools/toolchain/linalg.py", line 104, in _set_blas_variables
    raise EasyBuildError("_set_blas_variables: BLAS_LIB not set")
EasyBuildError: '_set_blas_variables: BLAS_LIB not set'
```

The problem was introduced with the changes in #2810 (included in EasyBuild v3.9.0), in particular the `self.toolchain.reset()` call that is done for every iteration (other than the first), which resets the toolchain variables (`self.toolchain.variables`).
This is done to ensure that the build environment gets set up correctly for every iteration, see https://github.com/easybuilders/easybuild-framework/pull/2810#issuecomment-473471868.

The underlying cause is this call in `LinAlg._set_blas_variables`:

```python
self.BLAS_LIB = self.variables.nappend('LIBBLAS', [x % self.BLAS_LIB_MAP for x in self.BLAS_LIB])
```

During the first iteration, the value for `self.BLAS_LIB` is `[]`, which basically results in:

```python
self.BLAS_LIB = self.variables.nappend('LIBBLAS', [])
```

`nappend` returns `None` in this case, resulting in `self.BLAS_LIB` being set to `None`.
This then backfires when `LinAlg._set_blas_variables` gets called again after the reset...

When `self.BLAS_LIB` is set to `['']` (which is the current fix in #2914), then `self.BLAS_LIB` is set to a `LibraryList` instance after the `nappend` call which represents `['']`, effectively working around the bug.

I have a feeling that the actual problem is in `nappend`, but I can't figure out a proper fix without spending significantly more time on this, so the current fix in #2810 will have to do for now.

In the long term, we should replace the whole rats nest that is the logic that sets up the toolchain build environment, since it clearly deserves to die in a fire. That's a pretty big refactor though...

**Comment by boegel** [comment]
closing since #2914 just got merged...
